### PR TITLE
JIT policy eval and add T-Maze

### PR DESCRIPTION
**Pull request by taodav**
BIG PR ALERT!

We do a number of things in this PR. In descending order in terms of the size of the changes:

- `jit`ify the policy eval class. Things are super fast now! Especially for the `do_grad` function which needs to call `PolicyEval.run` a bunch. The biggest change is batchifying the functions in the class.
- Tests for policy eval and also for the analytical solution for value function learning. This is mainly done through the new `SimpleChain` environment, which is a 1-action 10-state MDP that is a chain that goes right to the next state, until it terminates at the last node with a reward of +1.
- Fixes for T-Maze. Most notably @dsctt mentioned that there needs to be a special observation for the terminal state - this is indeed very much true, and has been rectified for the T-Maze environment.
- Minor tweaks for the `MDP` class, to make things fast as well.

NOTE: I've been changing things so that `numpy` is `np` and `jax.numpy` is `jnp`.

<s>So all-in-all, our tests went from wall-clock time 83 seconds to 4 seconds on my local machine 🙂 </s>
edit: I'm an idiot and commented out the two grad tests
edit 2: tests went from 83s to 6s. I made sure to check all the tests are passing. Now analytical `do_grad` is much faster!

### CODE REVIEW

File: grl/environment/tmaze_lib.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    """
+    n_states = 2 * n # Corridor
+    n_states += 2 # Start
+    n_states += 2 # Junction
+    n_states += 1 # Terminal state (includes both goal/non-goal)
+
+    T_up = np.eye(n_states, n_states)
+    T_down = T_up.copy()
+    T_up[[-1, -2, -3], [-1, -2, -3]] = 0
+    T_down[[-1, -2, -3], [-1, -2, -3]] = 0
```
Comment (camall3n) [inline grl/environment/tmaze_lib.py:23]: What's do these indices mean?

File: grl/environment/tmaze_lib.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    # If we go up or down at the junctions, we terminate
+    T_up[[-1 - 1, -2 - 1], [-1, -1]] = 1
+    T_down[[-1 - 1, -2 - 1], [-1, -1]] = 1
```
Comment (camall3n) [inline grl/environment/tmaze_lib.py:27]: Simplify math expressions: `[-2, -3]`. Even better would be to set constants like `JUNCTION_GOAL_UP` and `JUNCTION_GOAL_DOWN`

File: tests/test_grad.py [hunk 3]
```diff
@@ -2,7 +2,8 @@
 
 from grl import load_spec, do_grad, RTOL
 
-def test_example_7_p():
```
Comment (camall3n) [inline tests/test_grad.py:5]: This renaming is probably going to prevent pytest from automatically discovering the tests

File: tests/test_jax_pe.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            denom = np.where(denom == 0, 1,
+                             denom) # Avoid divide by zero (there may be a better way)
+            R_contributions /= denom
+
+            # R_obs_obs[:,curr_ob,next_ob] = R_contributions.sum(2).sum(1)
+            R_obs_obs = R_obs_obs.at[:, curr_ob, next_ob].set(R_contributions.sum(2).sum(1))
+
+    return MDP(T_obs_obs, R_obs_obs, amdp.p0, amdp.gamma)
+
+def indv_spec_jaxify_pe_funcs(spec):
```
Comment (camall3n) [inline tests/test_jax_pe.py:74]: Can you add a 1-line comment saying what this function tests?

File: tests/test_tmaze.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        ob = pomdp.observe(s)
+        if s == 0:
+            assert ob == 0
+        elif s == 1:
+            assert ob == 1
+
+    assert np.all(
+        np.isclose(start_counts / n_samples, np.zeros_like(start_counts) + 0.5, atol=1e-1))
+
+    # Now we test bumping
```
Comment (camall3n) [inline tests/test_tmaze.py:29]: What is bumping? Like self-loops?

File: tests/test_tmaze.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
+import numpy as np
+
+from grl.mdp import MDP, AbstractMDP
+from grl.environment.tmaze_lib import tmaze
+
+def test_tmaze():
```
Comment (camall3n) [inline tests/test_tmaze.py:6]: I prefer to split large tests up so there's only like 1-3 asserts per test.

**Review by camall3n — CHANGES_REQUESTED** [review]
Overall looks good. I pushed a few small changes already and requested a couple minor tweaks

**Review by dsctt — APPROVED** [review]
Awesome stuff man, lgtm, just gotta fix that test

File: grl/environment/examples_lib.py [hunk 7]
```diff
@@ -626,6 +628,91 @@ def simple_chain(n: int = 10):
...
+    pi[-2, 0] = 2 / 3
+    pi[-2, 1] = 1 / 3
+
+    Pi_phi = [pi]
+
+    # memory policy is observations * memory bits (2) x n_actions
+    Pi_phi_x = [Pi_phi[0].repeat(2, axis=0)]
+    return to_dict(T, R, discount, p0, phi_fully_observable, Pi_phi, Pi_phi_x)
+
+def tmaze_5_two_thirds_up_almost_fully_observable():
```
Comment (camall3n) [inline grl/environment/examples_lib.py:662]: Could use a comment explaining what this means

File: grl/environment/memory_lib.py [hunk 8]
```diff
@@ -6,7 +7,22 @@
 Dimensions: AxZxMxM
 """
 
-mem_0 = np.array([
+def get_memory(memory_id: int, n_obs: int, n_actions: int, n_mem_states: int = 2) -> np.ndarray:
+    current_module = globals()
```
Comment (camall3n) [inline grl/environment/memory_lib.py:11]: Huh. Interesting. Never heard of `globals()`

File: grl/environment/memory_lib.py [hunk 9]
```diff
@@ -215,3 +231,59 @@
...
+        [0, 1],
+    ],
+    [ # terminal
+        [1, 0],
+        [0, 1],
+    ],
+])
+memory_16 = np.array([mem_16, mem_16, mem_16, mem_16]) # up, down, right, left
+
+# No memory for t-maze
```
Comment (camall3n) [inline grl/environment/memory_lib.py:262]: Is this comment still correct? What does it mean?

File: grl/environment/tmaze_lib.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    # If we go up or down at the junctions, we terminate
+    T_up[[JUNCTION_DOWN, JUNCTION_UP], [TERMINAL, TERMINAL]] = 1
+    T_down[[JUNCTION_DOWN, JUNCTION_UP], [TERMINAL, TERMINAL]] = 1
+
+    T_left = np.zeros((n_states, n_states))
+    T_right = T_left.copy()
+
+    # At the leftmost and rightmost states we transition to ourselves
+    T_left[[START_UP, START_DOWN], [START_UP, START_DOWN]] = 1
+    T_right[[JUNCTION_DOWN, JUNCTION_UP], [JUNCTION_DOWN, JUNCTION_UP]] = 1
```
Comment (camall3n) [inline grl/environment/tmaze_lib.py:40]: This is all much clearer, thank you!

File: grl/environment/tmaze_lib.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    R_up = R_left.copy()
+    R_down = R_up.copy()
+
+    # If rewarding state is north
+    R_up[JUNCTION_UP, TERMINAL] = 4
+    R_down[JUNCTION_UP, TERMINAL] = -0.1
+
+    # If rewarding state is south
+    R_up[JUNCTION_DOWN, TERMINAL] = -0.1
+    R_down[JUNCTION_DOWN, TERMINAL] = 4
```
Comment (camall3n) [inline grl/environment/tmaze_lib.py:64]: I'd suggest converting these scalars to function args with default values

File: grl/environment/tmaze_lib.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    phi[-3:-1, 3] = 1
+
+    # we have a special termination observations
+    phi[-1, 4] = 1
+
+    return T, R, discount, p0, phi
+
+def slippery_tmaze(n: int, discount: float = 0.9, slip_prob: float = 0.1):
+    T, R, discount, p0, phi = tmaze(n, discount=discount)
+    slip_T = np.eye(T.shape[-1] - 3) * slip_prob
```
Comment (camall3n) [inline grl/environment/tmaze_lib.py:94]: I'd love a comment about how slip is implemented. Like is it a uniform random action or a uniform random _non-selected_ action?

File: tests/test_grad.py [hunk 13]
```diff
@@ -67,4 +73,7 @@ def test_example_7_m():
     ])
     memory_grad = do_grad(spec, pi, 'm', lr=1)
 
-    assert np.allclose(memory_end, memory_grad, atol=1e-5)
+    assert np.allclose(memory_end, softmax(memory_grad, axis=-1), atol=1e-2)
```
Comment (camall3n) [inline tests/test_grad.py:76]: I'm a little nervous about using such a large tolerance. Are we sure it's doing what we want if we allow it to be off by 1e-2? Some of these value functions vary by smaller amounts than that.

**Review by camall3n — CHANGES_REQUESTED** [review]
A couple minor things, comments etc.

File: grl/environment/memory_lib.py (hunk 8 above)
Comment (taodav) [reply grl/environment/memory_lib.py:11]: Was trying to find a way to search through all the global names within the `memory_lib` module. Open to alternatives too!

File: grl/environment/examples_lib.py (hunk 7 above)
Comment (taodav) [reply grl/environment/examples_lib.py:662]: I'll just take this out, not really necessary.

Comment (taodav) [reply grl/environment/examples_lib.py:662]: It's to test out t-maze, except you always know where the reward is, but all corridor observations are aliased. It was mainly to see if in the "optimal 1-bit memory" case for t-maze we still have a lambda discrepancy due to the corridor aliasing (we do).

File: tests/test_grad.py (hunk 13 above)
Comment (taodav) [reply tests/test_grad.py:76]: For value functions, I would totally agree - yes, a tolerance of 1e-2 is too large, seeing as the scale of the returns will depend on reward sparsity and discount rate.

But since we're essentially comparing two probability distributions which are normalized between 0 and 1, I'm more comfortable with this tolerance.

I've also looked into trying to make this tolerance smaller, and the approaches I've tried haven't done what I expected them to. I've tried low-temperature softmax, scaling the learning rate, scaling the loss - all seem to result in similar tolerances. It seems that the softmax function is quite stubborn in terms of getting large enough parameters for a small tolerance.

File: grl/environment/examples_lib.py (hunk 7 above)
Comment (taodav) [reply grl/environment/examples_lib.py:662]: I've added a comment to explain what this is.

File: grl/environment/memory_lib.py (hunk 8 above)
Comment (camall3n) [reply grl/environment/memory_lib.py:11]: This is fine for now. It seems like maybe it's an indication that we should find a better way of organizing things though.

File: tests/test_grad.py (hunk 13 above)
Comment (camall3n) [reply tests/test_grad.py:76]: Fair enough. I just didn't want it to end up biting us if we thought it was converged because the test passed, but really it's off by some significant amount that's still within the large tolerance.

File: grl/environment/tmaze_lib.py (hunk 12 above)
Comment (camall3n) [reply grl/environment/tmaze_lib.py:101]: @taodav I still don't really understand this. I want to know, supposing I choose a particular action in a particular state, what is the probability of actually taking each of the actions, or alternatively the probability of actually ending up in each of the states.

I'm not seeing the connection between the code or the comment and the answer to this question.

**Review by camall3n — APPROVED** [review]
